crypto/rsa_extra/rsassa_pss_asn1.c

```diff
@@ -0,0 +1,317 @@
+  CBS null;
+  return (CBS_len(params) == 0) ||
+         (CBS_get_asn1(params, &null, CBS_ASN1_NULL) && (CBS_len(&null) == 0) &&
+          (CBS_len(params) == 0));
+}
+
+// Decode One-way Hash Functions: return one on success and zero on failure.
+// See 2.1. https://tools.ietf.org/html/rfc4055#page-5
+static int decode_one_way_hash(CBS *cbs, RSA_ALGOR_IDENTIFIER **hash_algor) {
+  CBS seq, oid;
+  if (CBS_get_asn1(cbs, &seq, CBS_ASN1_SEQUENCE) && (CBS_len(cbs) == 0) &&
+      CBS_get_asn1(&seq, &oid, CBS_ASN1_OBJECT) && is_absent_or_null(&seq) &&
+      parse_oid(&oid, rsa_pss_hash_functions,
+                OPENSSL_ARRAY_SIZE(rsa_pss_hash_functions), hash_algor)) {
+    return 1;
+  }
+  OPENSSL_PUT_ERROR(RSA, EVP_R_DECODE_ERROR);
+  return 0;
+}
+
+// Decode Mask Generation Functions: return one on success and zero on failure.
+// See 2.2. https://tools.ietf.org/html/rfc4055#page-5
+static int decode_mask_gen(CBS *cbs, RSA_MGA_IDENTIFIER **mga) {
+  CBS seq, mgf1_oid, hash_seq, hash_oid;
+  RSA_ALGOR_IDENTIFIER *mgf1 = NULL;
+  RSA_ALGOR_IDENTIFIER *hash_algor = NULL;
+  if (CBS_get_asn1(cbs, &seq, CBS_ASN1_SEQUENCE) && (CBS_len(cbs) == 0) &&
+      CBS_get_asn1(&seq, &mgf1_oid, CBS_ASN1_OBJECT) &&
+      CBS_get_asn1(&seq, &hash_seq, CBS_ASN1_SEQUENCE) &&
+      (CBS_len(&seq) == 0) &&
+      CBS_get_asn1(&hash_seq, &hash_oid, CBS_ASN1_OBJECT) &&
+      is_absent_or_null(&hash_seq) &&
+      parse_oid(&mgf1_oid, rsa_pss_mg_functions,
+                OPENSSL_ARRAY_SIZE(rsa_pss_mg_functions), &mgf1) &&
+      parse_oid(&hash_oid, rsa_pss_hash_functions,
+                OPENSSL_ARRAY_SIZE(rsa_pss_hash_functions), &hash_algor)) {
+    *mga = RSA_MGA_IDENTIFIER_new();
+    if (*mga != NULL) {
+      (*mga)->mask_gen = mgf1;
+      (*mga)->one_way_hash = hash_algor;
+      return 1;
+    }
+  }
+  OPENSSL_PUT_ERROR(RSA, EVP_R_DECODE_ERROR);
+  RSA_ALGOR_IDENTIFIER_free(mgf1);
+  RSA_ALGOR_IDENTIFIER_free(hash_algor);
+  return 0;
+}
+
+// get_context_specific_value sets |*out| to the contents of DER-encoded, ASN.1
+// element (not including tag and length bytes) and advances |seq| over it.
+// When the tag value does not exist, |seq| gets recovered.
+// It returns one when the element exists.
+static int get_context_specific_value(CBS *seq, CBS *out, int index) {
+  unsigned int tag_value = CBS_ASN1_CONTEXT_SPECIFIC | CBS_ASN1_CONSTRUCTED | index;
+  CBS seq_cp = {seq->data, seq->len};
+  if (CBS_get_asn1(seq, out, tag_value)) {
+    return 1;
+  } else {
+    // All fields of RSASSA-PSS-params can be absent, which means default.
+    // Recover |seq|.
+    seq->data = seq_cp.data;
+    seq->len = seq_cp.len;
+    return 0;
+  }
+}
+
+// Decode [0] HashAlgorithm of RSASSA-PSS-params: return one on success and zero
+// on failure. See 3.1. https://tools.ietf.org/html/rfc4055#page-7
+static int decode_pss_hash(CBS *seq, RSA_ALGOR_IDENTIFIER **hash_algor) {
+  CBS cs;
+  if (!get_context_specific_value(seq, &cs, 0)) {
```

Comment:
magic number: `0`.

True for all places where `get_context_specific_value` is used.

Reply:
The comment mentions `[0]`. For other places, `[1], [2] and [3]` are specified.  I will replace these with macro.

```
// Decode [0] HashAlgorithm of RSASSA-PSS-params: return one on success and zero
// on failure. See 3.1. https://tools.ietf.org/html/rfc4055#page-7
static int decode_pss_hash(CBS *seq, RSA_ALGOR_IDENTIFIER **hash_algor) {
  CBS cs;
  if (!get_context_specific_value(seq, &cs, 0)) {
```